File: app/src/main/java/com/kksionek/queuedroid/view/keyboard/KeyboardView.java
package com.kksionek.queuedroid.view.keyboard;

import android.content.Context;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.kksionek.queuedroid.R;
import com.kksionek.queuedroid.model.Settings;
import com.kksionek.queuedroid.model.keyboard.KeyboardViewAdapter;

public class KeyboardView extends LinearLayout {
    private static final String TAG = "KEYBOARDVIEW";

    private static final int MAX_LENGTH = 10;

    private final TextView mCurPointsTextView;
    private final RecyclerView mButtonRecylerView;
    private int mColsNum;

    public KeyboardView(Context context, AttributeSet attrs) {
        super(context, attrs);
        LayoutInflater.from(getContext()).inflate(R.layout.keyboard, this);

        mCurPointsTextView = (TextView) findViewById(R.id.cur_points);
        Button clearButton = (Button) findViewById(R.id.clear_button);
        clearButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                clearPoints();
            }
        });
        Button backspaceButton = (Button) findViewById(R.id.backspace_button);
        backspaceButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                CharSequence text = mCurPointsTextView.getText();
                if (text.length() == 1)
                    clearPoints();
                else
                    mCurPointsTextView.setText(text.subSequence(0, text.length() - 1));
            }
        });

        mButtonRecylerView = (RecyclerView) findViewById(R.id.button_grid_view);
        KeyboardViewAdapter keyboardViewAdapter = new KeyboardViewAdapter(getContext());
        keyboardViewAdapter.setOnKeyboardItemClickListener(new KeyboardListener());
        mButtonRecylerView.setAdapter(keyboardViewAdapter);
        if (isInEditMode())
            mColsNum = 5;
        else
            mColsNum = Settings.getKeyboardColumnsCount(getContext());
        mButtonRecylerView.setLayoutManager(
                new GridLayoutManager(
                        getContext(),
                        mColsNum,
                        GridLayoutManager.VERTICAL,
                        false));
        mButtonRecylerView.addItemDecoration(
                new SpacesItemDecoration(getContext(), R.dimen.keyboard_item_gap_size));
    }

    public void setColumnCount(int columnCount) {
        if (mColsNum != columnCount) {
            mColsNum = columnCount;
            mButtonRecylerView.setLayoutManager(new GridLayoutManager(
                    getContext(),
                    mColsNum,
                    GridLayoutManager.VERTICAL,
                    false));
        }
    }

    public int getPoints() {
        return Integer.valueOf(mCurPointsTextView.getText().toString());
    }

    public void clearPoints() {
        mCurPointsTextView.setText("0");
    }

    class KeyboardListener implements OnKeyboardItemClickListener {

        public void onClick(int position) {
            CharSequence text = mCurPointsTextView.getText();
            if (text.charAt(0) == '0')
                mCurPointsTextView.setText(String.valueOf(position));
            else if (text.length() < MAX_LENGTH)
                mCurPointsTextView.setText(text + String.valueOf(position));
            else
                Toast.makeText(getContext(), R.string.view_keyboard_too_long_input_message, Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/kksionek/queuedroid/view/keyboard/KeyboardView.kt
package com.kksionek.queuedroid.view.keyboard

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.LinearLayout
import android.widget.Toast
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.GridLayoutManager
import com.kksionek.queuedroid.R
import com.kksionek.queuedroid.databinding.KeyboardBinding
import com.kksionek.queuedroid.model.SettingsProviderImpl
import com.kksionek.queuedroid.model.keyboard.KeyboardViewAdapter

class KeyboardView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : LinearLayout(context, attrs) {

    private var keyboardColsNum = 0

    private val binding: KeyboardBinding

    fun setColumnCount(columnCount: Int) {
        if (keyboardColsNum != columnCount) {
            keyboardColsNum = columnCount
            binding.buttonGridView.layoutManager = GridLayoutManager(
                context,
                keyboardColsNum,
                GridLayoutManager.VERTICAL,
                false
            )
        }
    }

    val points: Int
        get() = binding.curPoints.text.toString().toInt()

    fun clearPoints() {
        binding.curPoints.text = "0"
    }

    inner class KeyboardListener : OnKeyboardItemClickListener {
        @SuppressLint("SetTextI18n")
        override fun onClick(position: Int) {
            val text = binding.curPoints.text.toString()
            when {
                text == "0" -> binding.curPoints.text = position.toString()
                text.length < MAX_LENGTH -> binding.curPoints.text = "$text$position"
                else -> Toast.makeText(
                    context, R.string.view_keyboard_too_long_input_message, Toast.LENGTH_SHORT
                ).show()
            }
        }
    }

    companion object {
        private const val TAG = "KEYBOARDVIEW"
        private const val MAX_LENGTH = 10
    }

    init {
        binding = KeyboardBinding.inflate(LayoutInflater.from(context), this, true).apply {
            clearButton.setOnClickListener { clearPoints() }
            backspaceButton.setOnClickListener {
                val text = curPoints.text
                if (text.length == 1) {
                    clearPoints()
                } else {
                    curPoints.text = text.subSequence(0, text.length - 1)
                }
            }
            buttonGridView.adapter = KeyboardViewAdapter(KeyboardListener())

            keyboardColsNum = if (isInEditMode) {
                5
            } else {
                SettingsProviderImpl(PreferenceManager.getDefaultSharedPreferences(context))
                    .getKeyboardColumnsCount()
            }
            buttonGridView.layoutManager = GridLayoutManager(
                context,
                keyboardColsNum,
                GridLayoutManager.VERTICAL,
                false
            )
            buttonGridView.addItemDecoration(
                SpacesItemDecoration(context, R.dimen.keyboard_item_gap_size)
            )
        }
    }
}